Drop custom file upload plugin in favor of CSS solution

MartijnCuppens opened the pull request:
CSS only alternative solution for https://github.com/twbs/bootstrap/pull/31085.
Advantages:
- No js dependencies
- No need to add translated button labels
- Labels are automatically adjusted when with `<input type="file" multiple>` 

TODO:
- [x] ~Wait until legacy Edge is dropped~ This isn't really necessary, since legacy Edge already has a decent fallback   
- [x] Extend migration guide
- [x] Check if the alpha version is correct
- [x] Fix border radii on iOS

https://deploy-preview-31955--twbs-bootstrap.netlify.app/docs/5.0/forms/form-control/#file-input
https://deploy-preview-31955--twbs-bootstrap.netlify.app/docs/5.0/forms/validation/#supported-elements

Closes #30265
Closes #31085

MartijnCuppens committed:
Drop custom file upload plugin in favor of CSS

mdo commented on site/content/docs/5.0/migration.md:
```diff
@@ -250,7 +250,7 @@ Changes to Reboot, typography, tables, and more.
   - Refactored `.form-file` markup to resolve some visual bugs while allowing translation and button text changes via HTML instead of CSS.
   - Dropped native `.form-control-file` and `.form-control-range` components entirely.
-  - Renamed `.custom-file` to `.form-file` (including variables).
+  - Renamed `.custom-file` to `.form-file` (including variables). **Watch out:** `.form-file` was later on dropped in `v5.0.0-alpha4`, now you can use `.form-control`.
```
```suggestion
  - Renamed `.custom-file` to `.form-file` (including variables). **Watch out:** `.form-file` was later on dropped in `v5.0.0-alpha3`, now you can use `.form-control`.
```

MartijnCuppens replied:
Yeah, I didn't know how much prio this had. We should also wait for legacy Edge support to be dropped, so I used alpha4. I'll add this as a TODO.

mdo commented:
Thoughts on adding `cursor: pointer` here? I know it's not required, but compared to `.form-control` on textual inputs, where the cursor changes to indicate interactivity, there's no change here.

patrickhlauke commented:
> Thoughts on adding `cursor: pointer` here? I know it's not required, but compared to `.form-control` on textual inputs, where the cursor changes to indicate interactivity, there's no change here.

this old chestnut again ;)

as we do it on buttons by default, sure. also worth adding the `<input type="color">` ?

MartijnCuppens commented:
> Thoughts on adding `cursor: pointer` here?

Fine by me. Only on the button or the whole input?

patrickhlauke commented:
everything that's clickable, so the whole thing i'd say

patrickhlauke commented:
also, good work on this btw. looks/works nicely

MartijnCuppens commented:
Ok, should we add hover states? If yes, what kind of hoverstate? Shading the button background color maybe?

MartijnCuppens commented:
I've added a subtle hover state

mdo commented:
@MartijnCuppens this is amazing. Love the little hover state.

alecpl commented:
There's some padding issue in Firefox 82 on Ubuntu.
![input-bug](https://user-images.githubusercontent.com/546788/97100384-2ca5d080-1693-11eb-8d10-2946f62a1197.png)

MartijnCuppens committed:
Fix Firefox padding issue

Caused by https://searchfox.org/mozilla-central/rev/0b7007a23bc16c857f894140e12f307bfeef2fdd/layout/style/res/forms.css#494

MartijnCuppens commented:
@alecpl, 
Nice catch
That seems to be caused by these 5px:
https://searchfox.org/mozilla-central/rev/0b7007a23bc16c857f894140e12f307bfeef2fdd/layout/style/res/forms.css#494

I've now added a `-moz-margin-end` property to compensate these 5px.

patrickhlauke commented on scss/forms/_form-control.scss:
```diff
@@ -23,6 +23,14 @@
   @include box-shadow($input-box-shadow);
   @include transition($input-transition);
 
+  &[type="file"] {
```
while not related to this, mind if we slip in `&[type="color"]` here as well?

patrickhlauke replied:
otherwise, to keep it clean, happy to wait for this to be merged and then adding it in a new PR that's targeted

alecpl commented:
@MartijnCuppens The fix is not working. And I don't see how a "horizontal" margin-end could fix a "vertical" issue.

alecpl commented:
Maybe I wasn't clear about the issue description. I'm talking about the space (1 or 2px) between the button and bottom border.

MartijnCuppens commented:
@alecpl, the issue I fixed was the additional padding between the button and the input text, see this:
![image](https://user-images.githubusercontent.com/11559216/97157512-787e7580-1778-11eb-9e04-a4978d87819f.png)

I can't reproduce the issue you have over there on Mac or Windows. It's also rather small (I didn't spot it until you mentioned what was wrong), so I'm not sure this is a real deal breaker. Thoughts @twbs/team?

ffoodd commented:
I can see this on Ubuntu, only on Firefox. Feels weird, can't find anything yet. Seems related to `line-height` but I can't spot any reason for now…

alecpl commented:
This is related to font-family. If I remove "font-family:inherit" from the input the font will change to "Noto Sans" and the issue disappears.

MartijnCuppens commented:
Hmm, I can't really debug this myself here, if anyone can come up with a fix, please share 🙂

MartijnCuppens commented:
I zoomed in a bit and detected the error. Here, it was caused by the `min-height` rounding. Do you guys still have this issue when you remove `min-height` from the file input?

ffoodd commented:
@MartijnCuppens Doesn't seem to change anything, sadly. But maybe changing it in the devtools is not the best way?

alecpl commented:
I've found a fix: `.form-control::file-selector-button { font-family: inherit; }`. It can be also `font-family: unset`.

mdo reviewed (approved):
Love this PR! :D

mdo commented:
Also going to cross reference that #31993 fixes the issues @alecpl mentions here.

alecpl commented:
@mdo, #31993 does not fix my issue, need the `font-family:inherit` fix mentioned in my previous comment.

A few more words about the issue. Looks like firefox does not inherit font-family on the file input button from the input. It needs to be told to do so explicitely. Without the fix the computed font-family is "Noto Sans" on my computer. This font's line-height is 1px higher than what I get when I add the inherit rule.

ffoodd commented:
Very nice finding @alecpl, I confirm it solves the issue. 

We should find some reference to document this behaviour, either a bugzilla link or something… The [MDN page for `::file-selector-button` does not mention this](https://developer.mozilla.org/en-US/docs/Web/CSS/::file-selector-button) but made me notice the need for `::-webkit-file-upload-button` version too: does Autoprefixer handles this one?

- [Current spec discussion](https://github.com/w3c/csswg-drafts/issues/5049)
- **(!)** [Bugzilla updated 23 days ago](https://bugzilla.mozilla.org/show_bug.cgi?id=1635675)
- My findings: I guess it relates to [how the font is set](https://hg.mozilla.org/mozilla-central/rev/ff6c38ffa425#l7.104): see how the pseudo element gets `-moz-button` font-family directly?) however it should apply to every [Firefox > 82 out there, which rolled out only a few days ago](https://www.mozilla.org/en-US/firefox/82.0/releasenotes/). I'll cross-ref a link to bugzilla later.

XhmikosR commented:
We should ask upstream if it's not handled. And I'll try to move forward
with the postcss update in the next days.

On Wed, Oct 28, 2020, 11:16 Gaël Poupard <notifications@github.com> wrote:

> Very nice finding @alecpl <https://github.com/alecpl>, I confirm it
> solves the issue.
>
> We should find some reference to document this behaviour, either a
> bugzilla link or something… The MDN page for ::file-selector-button does
> not mention this
> <https://developer.mozilla.org/en-US/docs/Web/CSS/::file-selector-button>
> but made me notice the need for ::-webkit-file-upload-button version too:
> does Autoprefixer handles this one?
>
>    - Current spec discussion
>    <https://github.com/w3c/csswg-drafts/issues/5049>
>    - *(!)* Bugzilla updated 23 days ago
>    <https://bugzilla.mozilla.org/show_bug.cgi?id=1635675>
>    - My findings: I guess it relates to how the font is set
>    <https://hg.mozilla.org/mozilla-central/rev/ff6c38ffa425#l7.104>: see
>    how the pseudo element gets -moz-button font-family directly?) however
>    it should apply to every Firefox > 82 out there, which rolled out only
>    a few days ago
>    <https://www.mozilla.org/en-US/firefox/82.0/releasenotes/>. I'll
>    cross-ref a link to bugzilla later.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/twbs/bootstrap/pull/31955#issuecomment-717802443>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACVLNMUIVE45F2Q5W7WE7TSM7OODANCNFSM4S3GQEBA>
> .
>

MartijnCuppens committed:
Use the same font on the button as on the field itself

ffoodd commented:
FWIW the [bug I filled against Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1673819) got a very clear explanation (mainly, this pseudo-element should look like a button like it always did, not like an input).

I guess this "fix" @alecpl found could move to our rebot, and maybe suggested to Normalize and friends.

MartijnCuppens commented:
> We should find some reference to document this behaviour

It's shadow DOM, so all properties are reset to the default behaviour. So basically we should keep in mind these things need to re-reset (font size seems to be inherit by default though): 
https://github.com/twbs/bootstrap/blob/3e2f9ab237e0e338fbe497213ad2353e59fbee9c/scss/_reboot.scss#L431-L434

#31993 seems to be unrelated, for the issue @alecpl reported, but does seem to fix issues with errors in rounding when zoomed in/out which resulted in the same issue visually.

Thanks for the fix @alecpl, this was indeed needed, not only for FF, but also for Chrome (see https://codepen.io/MartijnCuppens/pen/yLJpmvE?editors=1100)

MartijnCuppens commented:
Hmm yeah, maybe we should indeed move this to reboot, I'll fix that

MartijnCuppens commented:
Ok, fixed the font-family (and all other font related properties) inheritance and also reported the 5px thing with Firefox, so we might be able to remove that dirty fix we now have in the future (https://bugzilla.mozilla.org/show_bug.cgi?id=1673895).

ffoodd commented:
> It's shadow DOM, so all properties are reset to the default behaviour.

Not really, as you may see in Firefox sources I linked earlier: https://hg.mozilla.org/mozilla-central/rev/ff6c38ffa425#l7.104 — only a few properties are set (and for example `::placeholder` does not require anything to inherit the font).

But that was fun to search, and a very nice PR! Great job @MartijnCuppens :100:

ffoodd reviewed (approved):
LGTM!

mdo reviewed (approved):
:shipit: when ready!

MartijnCuppens commented:
Shipped 🚢

I'll keep an eye on https://bugzilla.mozilla.org/show_bug.cgi?id=1673895 if FF is planning to change how they set their spacing

Johann-S commented:
That's amazing @MartijnCuppens nice work 👍
